extensions/indexes/lucene/src/test/xquery/lucene/facets.xql

```diff
@@ -0,0 +1,310 @@
+    let $from := ft:facets(head($result), "from", 10)
+    let $to := ft:facets(head($result), "to", 10)
+    return (
+        $where?Berlin, $where?Hamburg, $from?Susi, $to?Egon
+    )
+};
+
+declare
+    %test:arg("from", "Rudi")
+    %test:assertEquals(1)
+    %test:arg("from", "Susi")
+    %test:assertEquals(0)
+    %test:arg("from", "Rudi", "Hans")
+    %test:assertEquals(2)
+function facet:query-and-drill-down($from as xs:string+) {
+    let $options := map {
+        "facets": map {
+            "from": $from
+        }
+    }
+    return
+        count(collection("/db/lucenetest")//letter[ft:query(., "Berlin", $options)])
+};
+
+declare
+    %test:assertEquals(4, 2)
+function facet:store-and-remove() {
+    let $stored := xmldb:store("/db/lucenetest", "test2.xml", $facet:XML)
+    let $result := collection("/db/lucenetest")//letter[ft:query(., ())]
+    let $where := ft:facets(head($result), "place", 10)
+    return
+        $where?Berlin,
+    xmldb:remove("/db/lucenetest", "test2.xml"),
+    let $result := collection("/db/lucenetest")//letter[ft:query(., ())]
+    let $where := ft:facets(head($result), "place", 10)
+    return
+        $where?Berlin
+};
+
+declare
+    %test:arg("paths", "2017")
+    %test:assertEquals(2)
+    %test:arg("paths", "2019")
+    %test:assertEquals(2)
+    %test:arg("paths", "2019", "03")
+    %test:assertEquals(1)
+    %test:arg("paths", "2019", "03", "14")
+    %test:assertEquals(1)
+    %test:arg("paths", "2019", "03", "08")
+    %test:assertEquals(0)
+function facet:hierarchical-facets-query($paths as xs:string+) {
+    let $options := map {
+            "facets": map {
+                "date": $paths
+            }
+        }
+    let $result := collection("/db/lucenetest")//letter[ft:query(., (), $options)]
+    return
+        count($result)
+};
+
+declare
+    %test:arg("paths", "2017")
+    %test:assertEquals('{"03":2}')
+    %test:arg("paths", "2019")
+    %test:assertEquals('{"03":1,"04":1}')
+function facet:hierarchical-facets-retrieve($paths as xs:string*) {
+    let $result := collection("/db/lucenetest")//letter[ft:query(., ())]
+    let $facets := ft:facets(head($result), "date", 10, $paths)
+    return
+        serialize($facets, map { "method": "json" })
+};
+
+declare
+    %test:assertEquals('{"Berlin":2,"Hamburg":1}','{"Wrocław":2}')
+function facet:hierarchical-place() {
+    let $result := collection("/db/lucenetest")//letter[ft:query(., ())]
+    let $facets := ft:facets(head($result), "location", 10) (: Returns facet counts for "Germany" and "Poland" :)
+    for $country in map:keys($facets)
+    order by $country
+    return
+        serialize(
+            ft:facets(head($result), "location", 10, $country), (: Get facet counts for sub-categories :)
+            map { "method": "json" }
+        )
+};
+
+declare
+    %test:args("place:hamburg")
+    %test:assertEquals(1)
+    %test:args("place:berlin")
+    %test:assertEquals(2)
+    %test:args("from:rudi and place:berlin")
```

Comment:
For this and the next few argument sets, the boolean `AND` should be expressed in all caps. Otherwise, it is likely dropped as a stop word and thus fails to perform the expected function. See https://lucene.apache.org/core/2_9_4/queryparsersyntax.html#Boolean%20operators.